### PR TITLE
Buttons

### PR DESCRIPTION
**Pull request by Skillz4Killz**
This PR adds Buttons functionality to the library as well as minor fixes/cleanup.

Limits: 5 rows of 5 buttons i.e. max 25 buttons compared to max 20 reactions
image

Examples:
image

### CODE REVIEW

File: src/helpers/messages/send_message.ts [hunk 1]
```diff
@@ -93,6 +96,60 @@ export async function sendMessage(
     }
   }
 
+  if (content.components?.length) {
+    let actionRowCounter = 0;
+
+    for (const component of content.components) {
+      // 5 Link buttons can not have a custom_id
```
Comment (itohatweb) [inline src/helpers/messages/send_message.ts:103]:
```suggestion
      // 5 Link buttons can not have a customId
```

File: src/helpers/messages/send_message.ts [hunk 2]
```diff
@@ -93,6 +96,60 @@ export async function sendMessage(
...
+        ) {
+          throw new Error(Errors.LINK_BUTTON_CANNOT_HAVE_CUSTOM_ID);
+        }
+        if (
+          !component.customId
+        ) {
+          if (component.type !== ButtonStyles.Link) {
+            throw new Error(Errors.BUTTON_REQUIRES_CUSTOM_ID);
+          }
+        }
```
Comment (itohatweb) [inline src/helpers/messages/send_message.ts:117]: Little bit less nesting
```suggestion
        if (component.type === ButtonStyles.Link) {
          if (component.customId) throw new Error(Errors.LINK_BUTTON_CANNOT_HAVE_CUSTOM_ID);
        } else if (!component.customId) throw new Error(Errors.BUTTON_REQUIRES_CUSTOM_ID);
```

File: src/types/messages/components/action_row.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import { SnakeCasedPropertiesDeep } from "../../util.ts";
+import { ButtonComponent } from "./button_component.ts";
+
+export interface ActionRow {
```
Comment (itohatweb) [inline src/types/messages/components/action_row.ts:4]:
```suggestion
// TODO: add docs link
export interface ActionRow {
```

File: src/types/messages/components/action_row.ts [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+  /** Action rows are a group of buttons. */
+  type: 1;
+  /** The button components */
+  components: ButtonComponent[];
+}
+
+export type DiscordActionRow = SnakeCasedPropertiesDeep<ActionRow>;
```
Comment (itohatweb) [inline src/types/messages/components/action_row.ts:11]:
```suggestion
```

File: src/types/messages/components/button_component.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
+import { SnakeCasedPropertiesDeep } from "../../util.ts";
+import { ButtonStyles } from "./button_styles.ts";
+
+export interface ButtonComponent {
```
Comment (itohatweb) [inline src/types/messages/components/button_component.ts:4]:
```suggestion
// TODO: add docs link
export interface ButtonComponent {
```

File: src/types/messages/components/button_styles.ts [hunk 6]
```diff
@@ -0,0 +1,12 @@
+export enum ButtonStyles {
```
Comment (itohatweb) [inline src/types/messages/components/button_styles.ts:1]:
```suggestion
// TODO: add docs link
export enum DiscordButtonStyles {
```
Note: Add an alias named ButtonStyles at the bottom

File: src/types/messages/components/message_component_types.ts [hunk 7]
```diff
@@ -0,0 +1,6 @@
+export enum MessageComponentTypes {
```
Comment (itohatweb) [inline src/types/messages/components/message_component_types.ts:1]:
```suggestion
// TODO: add docs link
export enum DiscordMessageComponentTypes {
```
Note: Add an alias named MessageComponentTypes at the bottom

File: src/types/messages/message.ts [hunk 8]
```diff
@@ -82,4 +83,6 @@ export interface Message {
   referencedMessage?: Message;
   /** Sent if the message is a response to an Interaction */
   interaction?: MessageInteraction;
+  /** The components related to this message */
+  components: MessageComponents;
```
Comment (itohatweb) [inline src/types/messages/message.ts:87]: Is it always defined?

**Review by itohatweb — CHANGES_REQUESTED** [review]

File: src/types/messages/message.ts (hunk 8 above)
Comment (Skillz4Killz) [reply src/types/messages/message.ts:87]: There is no official docs on this yet so not 100% but yes i believe itll b an empty array

File: src/types/interactions/interaction.ts [hunk 9]
```diff
@@ -25,4 +26,6 @@ export interface Interaction {
   token: string;
   /** Read-only property, always `1` */
   version: 1;
+  /** for the message the button was attached to */
```
Comment (itohatweb) [inline src/types/interactions/interaction.ts:29]:
```suggestion
  /** For the message the button was attached to */
```

**Review by itohatweb — APPROVED** [review]

**Pull request merged**